underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java

```diff
@@ -0,0 +1,520 @@
+  }
+
+  // Not supported
+  @Override
+  public List<String> getFileLocations(String path) throws IOException {
+    return null;
+  }
+
+  // Not supported
+  @Override
+  public List<String> getFileLocations(String path, long offset) throws IOException {
+    return null;
+  }
+
+  @Override
+  public long getFileSize(String path) throws IOException {
+    StorageObject details = getObjectDetails(path);
+    if (details != null) {
+      return details.getContentLength();
+    } else {
+      throw new FileNotFoundException(path);
+    }
+  }
+
+  @Override
+  public long getModificationTimeMs(String path) throws IOException {
+    StorageObject details = getObjectDetails(path);
+    if (details != null) {
+      return details.getLastModifiedDate().getTime();
+    } else {
+      throw new FileNotFoundException(path);
+    }
+  }
+
+  // This call is currently only used for the web ui, where a negative value implies unknown.
+  @Override
+  public long getSpace(String path, SpaceType type) throws IOException {
+    return -1;
+  }
+
+  @Override
+  public boolean isFile(String path) throws IOException {
+    return exists(path) && !isFolder(path);
+  }
+
+  @Override
+  public String[] list(String path) throws IOException {
+    // Non recursive list
+    path = path.endsWith(PATH_SEPARATOR) ? path : path + PATH_SEPARATOR;
+    return listInternal(path, false);
+  }
+
+  @Override
+  public boolean mkdirs(String path, boolean createParent) throws IOException {
+    if (isFolder(path)) {
+      return true;
+    }
+    if (exists(path)) {
+      LOG.error("Cannot create directory " + path + " because it is already a file.");
+      return false;
+    }
+    if (!createParent) {
+      if (parentExists(path)) {
+        // Parent directory exists
+        return mkdirsInternal(path);
+      } else {
+        LOG.error("Cannot create directory " + path + " because parent does not exist");
+        return false;
+      }
+    }
+    // Parent directories should be created
+    if (parentExists(path)) {
+      // Parent directory exists
+      return mkdirsInternal(path);
+    } else {
+      String parentKey = getParentKey(path);
+      // Recursively make the parent folders
+      return mkdirs(parentKey, true) && mkdirsInternal(path);
+    }
+  }
+
+  @Override
+  public InputStream open(String path) throws IOException {
+    try {
+      path = stripPrefix(path);
+      return mClient.getObject(mBucketName, path).getDataInputStream();
+    } catch (ServiceException se) {
+      LOG.error("Failed to open file: " + path, se);
+      return null;
+    }
+  }
+
+  @Override
+  public boolean rename(String src, String dst) throws IOException {
+    if (!exists(src)) {
+      return false;
+    }
+    if (exists(dst)) {
+      if (!isFolder(dst)) {
+        LOG.error("Unable to rename " + src + " to " + dst + " because destination already "
+            + "exists as a file.");
+        return false;
+      }
+      String srcName = getKeyName(src);
```

Comment:
I am confused here. When isFolder(dst), why we should proceed? 
It seems here we copy src under folder dst in this case? Not sure this matches the semantics of rename in hdfs/gluster ufs.